src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs

```diff
@@ -326,12 +327,12 @@ public static bool TryParseValue(ref Utf8JsonReader reader, out JsonDocument doc
         /// </exception>
         public static JsonDocument ParseValue(ref Utf8JsonReader reader)
         {
-            bool ret = TryParseValue(ref reader, out JsonDocument document, shouldThrow: true);
+            bool ret = TryParseValue(ref reader, out JsonDocument? document, shouldThrow: true);
             Debug.Assert(ret, "TryParseValue returned false with shouldThrow: true.");
-            return document;
+            return document!;
```

Comment:
Instead of adding a bang here, `Debug.Assert(document != null);` before returning.

Reply:
I'm surprised that the compiler doesn't understand that ret is true, therefore document is not null.  But, yeah, I'd just add the assert.

Reply:
https://github.com/dotnet/roslyn/issues/34665

Reply:
FYI: calling method have [NotNullWhen(true)] for `TryParseValue(ref Utf8JsonReader reader, [NotNullWhen(true)] out JsonDocument? document, bool shouldThrow)`. Do you think asserting returning value is not enough?

Reply:
> Do you think asserting returning value is not enough?

It should be, but clearly the compiler isn't understanding the flow and giving you an error (which is why you had to use bang). See the issue @stephentoub pointed to.

When you say "asserting" what do you mean?

`Debug.Assert(document != null)` will be enough. Annotating the called method with `[NotNullWhen(true)]` wasn't.

Reply:
I meant `[NotNullWhen(true)]` means document will be not null when true returned and we are asserting returning value here which is enough and no need to  assert document. Compiler just doesn't understood such transferred context

Reply:
Sounds good. Thanks.